### ISSUE TITLE
vectorize stochastic metapopulation model

### ISSUE THREAD

**Issue by goldingn** [goldingn/pop#19]
1. make `pop:::expected(transfun, patch)` work in vectorized form, so that the single transfun is evaluated against a list of patches
2. make `pop:::stoch_prob()` and `pop:::stoch_rate()` work in vectorized mode by changing `n` and the parameters specification
3. make dispersal as efficient (vectorised) as possible too - binomial on probabilities of dispersal (as above), fed into a multinomial. the following is fairly quick:

``` r
rdispersal <- function(p, N) {
  disp <- matrix(0, length(N), length(N))
  for(i in seq_along(N)) {
    disp[i, ] <- rmultinom(1, N, p[i, ])
  }
  colSums(disp)
}
```

but could be improved by ignoring links with very low dispersal probabilities (ends of the tail) e.g. via an implementation of rmultinom that uses sparse matrices. Detecting that sparsity efficiently will be the more tricky aspect.

**Comment by goldingn** [comment]
see here for similar discussion: http://stackoverflow.com/questions/23097269/efficient-multinomial-sampling-when-sample-size-and-probability-vary

**Comment by goldingn** [comment]
This is slightly faster:

``` r
rdispersal <- function(N, p) {
  disp <- N * 0
  for (i in seq_along(N)) {
    disp <- disp + rmultinom(1, N[i], p[i, ])
  }
  disp
}
```

and faster than the fastest pure-R version in than SO thread (for this application). Stick with this for now.

**Comment by goldingn** [comment]
~ 160x faster to do vectorised rbinom than sequential for 5000-patch metapopulation, so well worth doing

**Comment by goldingn** [comment]
#30 now vectorizes across patches for within-patch dynamics

**Comment by goldingn** [comment]
closed in #30
